DetectorDescription/DDCMS/src/DDFilteredView.cc

```diff
@@ -614,62 +614,103 @@ const ExpandedNodes& DDFilteredView::history() {
   return nodes_;
 }
 
+/*
+ * Among all SepcPar XML sections defining the key,
+ * find the first SpecPar section containing a path matching the current node path.
+ */
 const DDSpecPar* DDFilteredView::find(const std::string& key) const {
-  DDSpecParRefs refs;
-  registry_->filter(refs, key, "");
-  int level = it_.back().GetLevel();
-  for (auto const& i : refs) {
-    auto k = find_if(begin(i->paths), end(i->paths), [&](auto const& j) {
-      auto topos = j.size();
-      auto frompos = j.rfind('/');
-      bool flag = false;
-      for (int nit = level; frompos - 1 <= topos and nit > 0; --nit) {
-        std::string_view name = it_.back().GetNode(nit)->GetVolume()->GetName();
-        std::string_view refname{&j[frompos + 1], topos - frompos - 1};
-        topos = frompos;
-        frompos = j.substr(0, topos).rfind('/');
-
-        auto rpos = refname.find(':');
-        if (rpos == refname.npos) {
-          name = noNamespace(name);
-        } else {
-          if (name.find(':') == name.npos) {
-            refname.remove_prefix(rpos + 1);
-          }
-        }
-        auto cpos = refname.rfind('[');
-        if (cpos != refname.npos) {
-          if (std::stoi(std::string(refname.substr(cpos + 1, refname.rfind(']')))) == copyNum()) {
-            refname.remove_suffix(refname.size() - cpos);
-            flag = true;
-            continue;
-          } else {
-            flag = false;
-            break;
-          }
-        }
-        if (!dd4hep::dd::isRegex(refname)) {
-          if (!dd4hep::dd::compareEqual(name, refname)) {
-            flag = false;
-            break;
-          } else {
-            flag = true;
-            continue;
+  // Filter the XML SpecPar sections: only those containing key are kept.
+  DDSpecParRefs allSpecParXMLSections;
```

Comment:
@ghugo83 - thanks! Do you have a version without renaming everything? It's hard to compare the changes especially with other PRs that are about to be merged.

Reply:
@ianna Yep sorry, I did some cleaning on the way (did not touch performance).
The 2 fixes are:
- continue statement at : https://github.com/cms-sw/cmssw/pull/31677/files#diff-06d03fa40cda4df4cac28c118cada0eaL645
- use of copyNum() at : https://github.com/cms-sw/cmssw/pull/31677/files#diff-06d03fa40cda4df4cac28c118cada0eaL642

Reply:
Yes luckily it only affects` const DDSpecPar* DDFilteredView::find(const std::string& key) const`, which is barely touched in the other PR. Only `registry_->filter(refs, key, "");` is changed in the other PR, which is independent from the fixes of this PR.

Reply:
ok, thanks! I'll add the fixes in the pending PR, then you can rebase this one :-)

Reply:
Hmm but the other PR was about performance?

Reply:
Also it would be nice to keep all the cleaning done in this PR no?

Reply:
Hopefully the following fixes it: https://github.com/cms-sw/cmssw/pull/31662/files#diff-06d03fa40cda4df4cac28c118cada0eaL617-R685

Reply:
@ghugo83 - the method is way too long and is being refactored to use shorted helper functions (improving performance as well) ;-)

Reply:
@ianna Hmm ok, whatever you prefer. But now I need to remove my fixes from my PR?
Also now, there will be a conflict between the 2 PRs, while before there was only a one line overlap?

Reply:
@ianna Shall I close this PR? Since you integrated the fixes.
Or maybe I could leave it open, and keep the cleaning in the files other than FilteredView?
I think the cleaning here in FilteredView was very nice, allows to understand better what is done, but if you refactor this later on in the step 4 PR, it is gonna be too messy to have several PRs modifying the same portions of the code.

Reply:
> > recent IB timing for DD4hep ModuleNumbering producer is about 6x better :-)
> 
> Cool, but that's the performance PRs with the speedup points in FilteredView we discussed in another thread (plus more I presume)?
> 
> Here, this is focusing on getting the same results with DD4hep than old DD.
> On this topic, there is one last fix from the TO DO list, which needs to be done. It regards the namespaces.
> Basically, the namespaces, when specified in the SpecPars in XMLs, are kept.
> But the nodes names, as used in FilteredView, sometimes lose their namespace (and that is the tricky point: where?). A noNamespace function is applied on them anyway.
> **Hence the `accept()` filter (not the `find()`) within the FilteredView,** compare paths with namespace with paths with no namespace.
> This results in all the paths containing a namespace in XMLs to be ignored.

yes, but the main offenders according to the profiler are the `tbb` containers and regular expressions.

Reply:
> @ianna Shall I close this PR? Since you integrated the fixes.
> Or maybe I could leave it opened, and keep the cleaning in the files other than FilteredView?
> I think the cleaning here in FilteredView was very nice, allows to understand better what is done. But if you refactor this later on in the step 4 PR, it will probably be a too big rebasing loss of time for both of us, to have several PRs modifying the same portions of the code.
> Let me know :)

I think, if you remove DDFilteredView from this PR it should be fine to integrate the rest :-)

BTW, I'm not sure **nice** is the word I'd choose - it took me quite some time to review it. I had to compare it line by line to decipher the renaming from the actual bug fix changes. One of these lines introduced a performance penalty for constructing a vector that was not needed. I'm sorry, it was faster to edit the original code.

When we read our own code, we have a mental model of how things are connected and how they work. When faced with “someone else’s code” we don’t have that. We are faced with pages and pages of code. And often written in a style that is dissimilar to our own.

This can lead us to think that the style of the code is the problem. That if only the style was “correct” (read: like my own) then it would be “easier” to understand.

The fundamental problem with reading someone else’s code is **the lack of a mental model**.

One way to approach a code base is to code your way to understanding, by refactoring the code. I really recommend that you consider the actual refactoring as “throw-away”. It’s hard not to become attached to one’s refactoring, but I implore you to try.

There are many ways of doing this, ranging all the way from high level architectural changes to style guided refactorings. Both of these extremes are, however, a bit dangerous since they tend to make one a bit arrogant and might make one blind to underlying reasons for why things are the way they are. I would not recommend sharing your refactorings with the projects maintainers, that might start you off on the wrong foot :-)

Reply:
@ianna In the FilteredView, I mostly added comments and renamed variable, but I agree it is personal style.
This was on the fly of the actual bug fixes, it was to help me, and was not the actual purpose of this PR.

Personally, I dislike variables called name, rpos, cpos, topos, i, j, k, ref, or which just repeat the type of the object. They are faster to type, but I do not understand what they represent, which, to me, makes things way harder and error-prone to read.

I totally agree with you that it is a matter of personal style, no pb, I will just remove the changes in FilteredView :)